### PR TITLE
[NCL-5386] - Implementation of native query to retrieve minimized list of artifact…

### PR DESCRIPTION
**Pull request by vibe13**
… dependencies given a build record

### Checklist:

* [ ] Have you added a note in the [CHANGELOG wiki](https://github.com/project-ncl/pnc/wiki/Changelog) for your change if user-facing?
* [ ] Have you added unit tests for your change?

### CODE REVIEW

**Comment by vibe13** [comment]
retest this please

**Comment by vibe13** [comment]
retest this please

**Comment by vibe13** [comment]
retest this please

**Comment by vibe13** [comment]
retest this please

**Comment by vibe13** [comment]
@thescouser89 @jbartece Hi could you please review again this one? I could find another way to override at runtime some EAGER fetching (see https://projects.engineering.redhat.com/browse/NCL-5376), so I created an ad-hoc endpoint that calls a native query. Thanks!

**Comment by vibe13** [comment]
retest this please

**Comment by vibe13** [comment]
retest this please

File: rest-model/src/main/java/org/jboss/pnc/rest/restmodel/TargetRepositoryRest.java [hunk 1]
```diff
@@ -78,6 +78,15 @@
     @NotNull(groups = {WhenUpdating.class, WhenCreatingNew.class})
     private Set<Integer> artifactIds = new HashSet<>();
 
+    public TargetRepositoryRest(Integer id, Boolean temporaryRepo,
```
Comment (jbartece) [inline rest-model/src/main/java/org/jboss/pnc/rest/restmodel/TargetRepositoryRest.java]: I think, it is redundant as we have a Builder for this class.

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/internal/ArtifactSpringRepository.java [hunk 2]
```diff
@@ -18,12 +18,108 @@
...
+                    "FROM Artifact artifact " +
+                    "INNER JOIN build_record_built_artifact_map built_artifact " +
+                    "  ON built_artifact.built_artifact_id = artifact.id " +
+                    "INNER JOIN BuildRecord buildRecord " +
+                    "  ON built_artifact.build_record_id = buildRecord.id " +
+                    "INNER JOIN TargetRepository targetRepository " +
+                    "  ON targetRepository.id = artifact.targetRepository_id " +
+                    "WHERE " +
+                    " buildRecord.id = ?1",
+            nativeQuery = true)
```
Comment (jbartece) [inline datastore/src/main/java/org/jboss/pnc/datastore/repositories/internal/ArtifactSpringRepository.java]: What is the reason to set it as nativeQuery?

**Comment by vibe13** [comment]
> What are the reasons behind this PR? To improve performance by not loading dependent entities?
@jbartece yes as mentioned in the call earlier it's because the number of joins that gets triggered from the rest endpoints. Unfortunately with the current model, I could not override at runtime the EAGER fetching, so they query is extremely slow and returns data which are not needed by the Maven Generation zip.

> What is the reason behind the RawArtifact? Is it not to contain the deps?
Yes it does not contain any deps from Artifact to BuildRecords. Using native queries, it's more convenient to have an interface which can automatically map the Object[] returned to something meaningful, rather than doing that mapping yourself. I could not use the ArtifactRest class as rest-model is not meant to be visible from that package. And did not want to use Artifact entity class because the results of a native query get managed by Hibernate if mapped to entities, and did not want to incur in the problem of wiping out the collections of the returned Artifacts :-)
 
> Do you plan it to port it to 2.0 as well?
First I will check what kind of queries get generated from that REST endpoints, and would rather fix the mapping if it's the case. I would not like to port native queries to 2.0

File: rest-model/src/main/java/org/jboss/pnc/rest/restmodel/TargetRepositoryRest.java (hunk 1 above)
Comment (vibe13) [reply rest-model/src/main/java/org/jboss/pnc/rest/restmodel/TargetRepositoryRest.java]: Fair point, will fix that!

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/internal/ArtifactSpringRepository.java (hunk 2 above)
Comment (vibe13) [reply datastore/src/main/java/org/jboss/pnc/datastore/repositories/internal/ArtifactSpringRepository.java]: With standard JPA I could not mange to override at runtime some EAGER fetches, so I had to go to native queries.

**Comment by vibe13** [comment]
retest this please

**Pull request merged**